### PR TITLE
check invalid parameters

### PR DESCRIPTION
**Pull request by wxchan**
Give a warning when meet "invalid" parameters. Can help avoid typos.

### CODE REVIEW

File: include/LightGBM/config.h [hunk 1]
```diff
@@ -437,10 +438,39 @@ struct ParameterAlias {
...
+      "fair_c", "poission_max_delta_step", "scale_pos_weight",
+      "boost_from_average", "max_position", "label_gain",
+      "metric", "metric_freq", "time_out",
+      "gpu_platform_id", "gpu_device_id", "gpu_use_dp"
+    });
     std::unordered_map<std::string, std::string> tmp_map;
     for (const auto& pair : *params) {
       if (alias_table.count(pair.first) > 0) {
         tmp_map[alias_table[pair.first]] = pair.second;
+        params->erase(pair.first);
```
Comment (wxchan) [inline include/LightGBM/config.h]: `params->erase` causes segfault here, but it didn't cause segfault on my machine.

Comment (wxchan) [inline include/LightGBM/config.h]: the param set in cmd line will check twice in this case.

File: include/LightGBM/config.h [hunk 2]
```diff
@@ -437,10 +438,38 @@ struct ParameterAlias {
...
+      "metric", "metric_freq", "time_out",
+      "gpu_platform_id", "gpu_device_id", "gpu_use_dp",
+      "convert_model", "convert_model_language"
+    });
     std::unordered_map<std::string, std::string> tmp_map;
     for (const auto& pair : *params) {
       if (alias_table.count(pair.first) > 0) {
         tmp_map[alias_table[pair.first]] = pair.second;
+      } else if (parameter_set.count(pair.first) == 0) {
+        Log::Warning("Unknown parameter: %s", pair.first.c_str());
```
Comment (guolinke) [inline include/LightGBM/config.h]: Maybe throw a FATAL error is better ? many user set verbosity to 0 in R and python package.

Comment (wxchan) [inline include/LightGBM/config.h]: Ok, but I may miss some parameters in parameter_set. You might check that.

Comment (wxchan) [inline include/LightGBM/config.h]: Why commit 0e04c28 is passing now? Did `params->erase(pair.first);` cause errors?

Comment (guolinke) [inline include/LightGBM/config.h]: @wxchan I am not sure, we can rebuild 0e04c28 for more times, and see what happens.

**Comment by YuriyGuts** [comment]
My existing code that used `num_boost_round`, `feature_fraction_seed`, and `bagging_fraction_seed` stopped working when I installed a new version today, reporting these parameters as invalid. Although the documentation lists these parameters as acceptable. Is this a bug or should I change something?

**Comment by wxchan** [comment]
@YuriyGuts 

feature_fraction_seed: is a valid parameter, sorry I forgot adding it. 

bagging_fraction_seed: I search this keyword in repo, seem this is never used. It should be bagging_seed. But perhaps we can add an alias.

num_boost_round: do you mean num_boost_round in python? It's not used in C++.

You can rollback to commit b5e042d9d1392aec523b82585526e16ab4800fff for temp solution, sorry for your inconvenience. We will fix it tomorrow.

@guolinke I think maybe we should still use warning instead of fatal, many users would like to re-use  params dict from other packages. But this indeed helps user find some parameters they think it was used but actually not.

**Comment by wxchan** [comment]
@YuriyGuts The param check happens here https://github.com/wxchan/LightGBM/blob/d19b20b984fa78615bceddfe66135a1282976335/include/LightGBM/config.h#L441-L481.

You can add `feature_fraction_seed`, `bagging_fraction_seed` and `num_boost_round` to`parameter_set`, or change `Log::Fatal` to `Log::Warning` in line 472.

**Comment by YuriyGuts** [comment]
@wxchan Yes, I used these parameters, including `num_boost_round`, in Python code, like in the snippet below. Reverting to a previous commit helps, thanks. Please let me know if I can help in some way.

```
lgb_params = {
    'objective': 'binary',
    'metric': 'binary_logloss',
    'boosting': 'gbdt',
    'device': 'cpu',
    'feature_fraction': 0.5,
    'num_leaves': 64,
    'learning_rate': 0.03,
    'num_boost_round': 3000,
    'early_stopping_rounds': 5,
    'verbose': 1,
    'bagging_fraction_seed': RANDOM_SEED,
    'feature_fraction_seed': RANDOM_SEED,
}

model = lgb.train(lgb_params, ...)
```

**Comment by YuriyGuts** [comment]
@wxchan re: `bagging_fraction_seed`, this is my mistake indeed, I somehow misread the docs and used the name by analogy with `feature_fraction_seed`.

**Comment by guolinke** [comment]
@wxchan @YuriyGuts 
Will ```num_boost_round``` in lgb_params will be used in lgb.train ? 
I think it should be ```model = lgb.train(lgb_params, num_boost_round=3000)```

**Comment by YuriyGuts** [comment]
@guolinke My actual training code looked like this:

```
    model = lgb.train(
        lgb_params,
        lgb_data_train,
        valid_sets=[lgb_data_train, lgb_data_val],
        evals_result=evals_result,
        num_boost_round=lgb_params['num_boost_round'],
        early_stopping_rounds=lgb_params['early_stopping_rounds'],
        verbose_eval=False,
    )
```
The only reason I did that was to have all parameters in one place and to avoid human error when running many tuning experiments. I agree that one might argue `lgb_params` shouldn't contain `num_boost_round`, especially if the general design direction is to have one explicit way of doing every task.

**Comment by YuriyGuts** [comment]
I also see that XGBoost doesn't seem to use `num_boost_round` if it's passed in the parameter dictionary but not passed as a keyword argument in `xgb.train`, so I guess it's fair to consider it an invalid parameter for consistency reasons.

**Comment by wxchan** [comment]
@guolinke num_boost_round is not used in params, but I think it's not so wrong to put it in params, we might not make it a fatal.

**Pull request merged**